### PR TITLE
[hotfix] Address the performance issue of topk.

### PR DESCRIPTION
**Pull request by yzh119**
## Description
The sortpool module is time-consuming in SEAL model, this is because our in-efficient implementation of topk (we called argsort instead of topk API for backend agnostic), this PR fix this issue by replacing argsort with topk for PyTorch backend.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/readout.py [hunk 1]
```diff
@@ -542,33 +542,42 @@ def _topk_on(graph, typestr, feat, k, descending, sortby, ntype_or_etype):
...
+    feat_ = F.pad_packed_tensor(feat, batch_num_objs, fill_val, l_min=k)  # (batch_size, l, d)
+    if F.backend_name == 'pytorch' and sortby is not None:
+        import torch as th
+        keys = feat_[..., sortby]  # (batch_size, l)
+        topk_indices = keys.topk(k, -1, largest=descending)[1]  # (batch_size, k)
+        feat_ = F.pad_packed_tensor(feat, batch_num_objs, 0, l_min=k)
+        feat_ = feat_.view((batch_size * length), -1)
+        shift = th.arange(0, batch_size, device=feat_.device).view(batch_size, 1) * length
+        topk_indices_ = topk_indices + shift
+        return feat_[topk_indices_].view(batch_size, k, -1), topk_indices
```
Comment (jermainewang) [inline python/dgl/readout.py:554]: put the two branches into two functions `torch_topk` and `general_topk`.

Comment (yzh119) [reply python/dgl/readout.py:554]: i see

File: python/dgl/backend/pytorch/tensor.py [hunk 2]
```diff
@@ -227,32 +227,40 @@ def randint(shape, dtype, ctx, low, high):
...
+    index = th.ones(len(input), dtype=th.int64, device=device)
+    cum_lengths = th.cumsum(lengths, 0)
+    index[cum_lengths[:-1]] += (max_len - lengths[:-1])
+    index = th.cumsum(index, 0) - 1
+    x[index] = input
+    return x.view(batch_size, max_len, *old_shape[1:])
 
 def pack_padded_tensor(input, lengths):
-    batch_size, max_len = input.shape[:2]
+    _, max_len = input.shape[:2]
```
Comment (jermainewang) [inline python/dgl/backend/pytorch/tensor.py:251]: `max_len = input.shape[1]`

File: python/dgl/readout.py [hunk 3]
```diff
@@ -485,6 +485,31 @@ def broadcast_edges(graph, graph_feat, *, etype=None):
...
+    ----------
+    keys : Tensor
+        The key for ranking.
+    k : int
+        The :math:`k` in "top-:math:`k`".
+    descending : bool
+        Indicates whether to return the feature corresponding to largest or
+        smallest elements.
+    x : Tensor
+        The padded feature with shape (batch, max_len, *)
```
Comment (jermainewang) [inline python/dgl/readout.py:502]: doc the returns

File: python/dgl/readout.py [hunk 4]
```diff
@@ -542,8 +567,14 @@ def _topk_on(graph, typestr, feat, k, descending, sortby, ntype_or_etype):
 
     length = max(max(F.asnumpy(batch_num_objs)), k)
     fill_val = -float('inf') if descending else float('inf')
-    feat_ = F.pad_packed_tensor(feat, batch_num_objs, fill_val, l_min=k)
+    feat_ = F.pad_packed_tensor(feat, batch_num_objs, fill_val, l_min=k)  # (batch_size, l, d)
+
+    if F.backend_name == 'pytorch' and sortby is not None:
+        # PyTorch's implementation of top-K
+        keys = feat_[..., sortby]  # (batch_size, l)
+        return _topk_torch(keys, k, descending, feat_)
```
Comment (jermainewang) [inline python/dgl/readout.py:575]: It's better to not have `return` in the middle for lengthy function.
```python
if F.backend_name == 'pytorch' and sortby is not None:
    ...
    return ...
else:
    # Fallback to ...
    return ...
```

**Review by jermainewang — CHANGES_REQUESTED** [review]

**Comment by jermainewang** [comment]
Also, what's the performance difference before v.s. after?

**Comment by yzh119** [comment]
@jermainewang I run the following script before and after this pr:
```python
import dgl
import time
import torch as th
from dgl.nn import SortPooling

gs = []
for _ in range(128):
    gs.append(dgl.rand_graph(128, 2048))

g = dgl.batch(gs)
g = g.to(0)

x = th.rand(128 * 128, 64).to(0)
sortpool = SortPooling(k=10)

th.cuda.synchronize()
tic = time.time()
for i in range(10):
    y = sortpool(g, x)
th.cuda.synchronize()
toc = time.time()

print((toc - tic) / 10)
```

Before this PR: 0.017972564697265624s
After this PR: 0.0015792131423950195s

This PR accelerates topk by 10x in this setting.

**Review by jermainewang — APPROVED** [review]

**Pull request merged**